File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/RegisterPinManifestInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Base64;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.regex.Pattern;

/** Typed first-release builder for the consensus {@code RegisterPinManifest} instruction. */
public final class RegisterPinManifestInstruction implements InstructionTemplate {

  public static final String ACTION = "RegisterPinManifest";
  private static final int MAX_MANIFEST_PAYLOAD_BYTES = 512 * 1024;
  private static final int MAX_ALIAS_PROOF_BYTES = 1024 * 1024;
  private static final Pattern CANONICAL_HEX = Pattern.compile("^[0-9a-f]+$");
  private static final Set<String> MANDATORY_ARGUMENTS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  "action",
                  "manifest_payload_base64",
                  "submitted_epoch")));
  private static final Set<String> OPTIONAL_ARGUMENTS =
      Collections.unmodifiableSet(
          new HashSet<>(
              Arrays.asList(
                  "successor_of_hex", "alias.name", "alias.namespace", "alias.proof_hex")));

  private final String manifestPayloadBase64;
  private final long submittedEpoch;
  private final String successorOfHex;
  private final AliasBinding aliasBinding;
  private final Map<String, String> arguments;

  private RegisterPinManifestInstruction(final Builder builder) {
    manifestPayloadBase64 = builder.manifestPayloadBase64;
    submittedEpoch = builder.submittedEpoch;
    successorOfHex = builder.successorOfHex;
    aliasBinding = builder.aliasBinding;
    arguments = Collections.unmodifiableMap(builder.canonicalArguments());
  }

  public String manifestPayloadBase64() {
    return manifestPayloadBase64;
  }

  /** Returns a fresh copy of the canonical Norito manifest payload. */
  public byte[] manifestPayloadBytes() {
    return Base64.getDecoder().decode(manifestPayloadBase64);
  }

  public long submittedEpoch() {
    return submittedEpoch;
  }

  public String successorOfHex() {
    return successorOfHex;
  }

  public AliasBinding aliasBinding() {
    return aliasBinding;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.REGISTER;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static Builder builder() {
    return new Builder();
  }

  public static RegisterPinManifestInstruction fromArguments(final Map<String, String> arguments) {
    Objects.requireNonNull(arguments, "arguments");
    if (!ACTION.equals(arguments.get("action"))) {
      throw new IllegalArgumentException("Instruction argument 'action' must be " + ACTION);
    }
    for (final String key : arguments.keySet()) {
      if (!MANDATORY_ARGUMENTS.contains(key) && !OPTIONAL_ARGUMENTS.contains(key)) {
        throw new IllegalArgumentException("Unsupported RegisterPinManifest argument: " + key);
      }
    }
    if (!arguments.keySet().containsAll(MANDATORY_ARGUMENTS)) {
      throw new IllegalArgumentException("RegisterPinManifest arguments are missing required fields");
    }
    return builder()
        .setManifestPayloadBase64(require(arguments, "manifest_payload_base64"))
        .setSubmittedEpoch(requireLong(arguments, "submitted_epoch"))
        .setSuccessorOfHex(arguments.get("successor_of_hex"))
        .setAliasBinding(AliasBinding.fromArguments(arguments))
        .build();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof RegisterPinManifestInstruction)) {
      return false;
    }
    final RegisterPinManifestInstruction other = (RegisterPinManifestInstruction) obj;
    return submittedEpoch == other.submittedEpoch
        && Objects.equals(manifestPayloadBase64, other.manifestPayloadBase64)
        && Objects.equals(successorOfHex, other.successorOfHex)
        && Objects.equals(aliasBinding, other.aliasBinding);
  }

  @Override
  public int hashCode() {
    return Objects.hash(manifestPayloadBase64, submittedEpoch, successorOfHex, aliasBinding);
  }

  public static final class Builder {
    private String manifestPayloadBase64;
    private Long submittedEpoch;
    private String successorOfHex;
    private AliasBinding aliasBinding;

    private Builder() {}

    public Builder setManifestPayloadBase64(final String manifestPayloadBase64) {
      this.manifestPayloadBase64 = requireCanonicalManifestPayload(manifestPayloadBase64);
      return this;
    }

    public Builder setManifestPayload(final byte[] manifestPayload) {
      Objects.requireNonNull(manifestPayload, "manifestPayload");
      if (manifestPayload.length == 0 || manifestPayload.length > MAX_MANIFEST_PAYLOAD_BYTES) {
        throw new IllegalArgumentException(
            "manifestPayload must contain 1.." + MAX_MANIFEST_PAYLOAD_BYTES + " bytes");
      }
      this.manifestPayloadBase64 =
          Base64.getEncoder().encodeToString(Arrays.copyOf(manifestPayload, manifestPayload.length));
      return this;
    }

    public Builder setSubmittedEpoch(final long submittedEpoch) {
      if (submittedEpoch < 0) {
        throw new IllegalArgumentException("submittedEpoch must be non-negative");
      }
      this.submittedEpoch = submittedEpoch;
      return this;
    }

    public Builder setSuccessorOfHex(final String successorOfHex) {
      this.successorOfHex =
          successorOfHex == null
              ? null
              : requireNonzeroDigest(successorOfHex, "successorOfHex");
      return this;
    }

    public Builder setAliasBinding(final AliasBinding aliasBinding) {
      this.aliasBinding = aliasBinding;
      return this;
    }

    public RegisterPinManifestInstruction build() {
      if (manifestPayloadBase64 == null) {
        throw new IllegalStateException("manifestPayload must be set");
      }
      if (submittedEpoch == null) {
        throw new IllegalStateException("submittedEpoch must be set");
      }
      return new RegisterPinManifestInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> result = new LinkedHashMap<>();
      result.put("action", ACTION);
      result.put("manifest_payload_base64", manifestPayloadBase64);
      result.put("submitted_epoch", Long.toString(submittedEpoch));
      if (successorOfHex != null) {
        result.put("successor_of_hex", successorOfHex);
      }
      if (aliasBinding != null) {
        aliasBinding.appendArguments(result);
      }
      return result;
    }
  }

  /** Optional manifest alias binding. */
  public static final class AliasBinding {
    private final String name;
    private final String namespace;
    private final String proofHex;

    private AliasBinding(final Builder builder) {
      name = builder.name;
      namespace = builder.namespace;
      proofHex = builder.proofHex;
    }

    public String name() {
      return name;
    }

    public String namespace() {
      return namespace;
    }

    public String proofHex() {
      return proofHex;
    }

    private void appendArguments(final Map<String, String> target) {
      target.put("alias.name", name);
      target.put("alias.namespace", namespace);
      target.put("alias.proof_hex", proofHex);
    }

    static AliasBinding fromArguments(final Map<String, String> arguments) {
      return fromArguments(arguments, false);
    }

    static AliasBinding fromArguments(
        final Map<String, String> arguments, final boolean required) {
      int present = 0;
      if (arguments.containsKey("alias.name")) {
        present++;
      }
      if (arguments.containsKey("alias.namespace")) {
        present++;
      }
      if (arguments.containsKey("alias.proof_hex")) {
        present++;
      }
      if (present == 0) {
        if (required) {
          throw new IllegalArgumentException(
              "Alias binding requires alias.name, alias.namespace, and alias.proof_hex");
        }
        return null;
      }
      if (present != 3) {
        throw new IllegalArgumentException(
            "Alias binding requires alias.name, alias.namespace, and alias.proof_hex together");
      }
      return builder()
          .setName(require(arguments, "alias.name"))
          .setNamespace(require(arguments, "alias.namespace"))
          .setProofHex(require(arguments, "alias.proof_hex"))
          .build();
    }

    public static Builder builder() {
      return new Builder();
    }

    @Override
    public boolean equals(final Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof AliasBinding)) {
        return false;
      }
      final AliasBinding other = (AliasBinding) obj;
      return Objects.equals(name, other.name)
          && Objects.equals(namespace, other.namespace)
          && Objects.equals(proofHex, other.proofHex);
    }

    @Override
    public int hashCode() {
      return Objects.hash(name, namespace, proofHex);
    }

    public static final class Builder {
      private String name;
      private String namespace;
      private String proofHex;

      private Builder() {}

      public Builder setName(final String name) {
        this.name = requireAliasText(name, "alias.name");
        return this;
      }

      public Builder setNamespace(final String namespace) {
        this.namespace = requireAliasText(namespace, "alias.namespace");
        return this;
      }

      public Builder setProofHex(final String proofHex) {
        this.proofHex =
            requireCanonicalHex(proofHex, "alias.proofHex", null, MAX_ALIAS_PROOF_BYTES);
        return this;
      }

      public AliasBinding build() {
        if (name == null || namespace == null || proofHex == null) {
          throw new IllegalStateException("alias name, namespace, and proof must be set");
        }
        return new AliasBinding(this);
      }
    }
  }

  private static String requireCanonicalManifestPayload(final String value) {
    if (value == null || value.isEmpty()) {
      throw new IllegalArgumentException("manifestPayloadBase64 must not be empty");
    }
    final int maximumEncodedLength = ((MAX_MANIFEST_PAYLOAD_BYTES + 2) / 3) * 4;
    if (value.length() > maximumEncodedLength) {
      throw new IllegalArgumentException("manifestPayloadBase64 exceeds the manifest size limit");
    }
    final byte[] decoded;
    try {
      decoded = Base64.getDecoder().decode(value);
    } catch (final IllegalArgumentException ex) {
      throw new IllegalArgumentException("manifestPayloadBase64 must be valid base64", ex);
    }
    if (decoded.length == 0 || decoded.length > MAX_MANIFEST_PAYLOAD_BYTES) {
      throw new IllegalArgumentException("manifestPayloadBase64 decoded size is outside limits");
    }
    if (!Base64.getEncoder().encodeToString(decoded).equals(value)) {
      throw new IllegalArgumentException("manifestPayloadBase64 must use canonical padded base64");
    }
    return value;
  }

  private static String requireNonzeroDigest(final String value, final String field) {
    final String digest = requireCanonicalHex(value, field, 32, null);
    boolean nonzero = false;
    for (int index = 0; index < digest.length(); index++) {
      if (digest.charAt(index) != '0') {
        nonzero = true;
        break;
      }
    }
    if (!nonzero) {
      throw new IllegalArgumentException(field + " must not be the all-zero digest");
    }
    return digest;
  }

  private static String requireCanonicalHex(
      final String value,
      final String field,
      final Integer expectedBytes,
      final Integer maximumBytes) {
    if (value == null
        || value.isEmpty()
        || value.length() % 2 != 0
        || !CANONICAL_HEX.matcher(value).matches()) {
      throw new IllegalArgumentException(
          field + " must be canonical lowercase even-length hex without a prefix");
    }
    if (expectedBytes != null && value.length() != expectedBytes * 2) {
      throw new IllegalArgumentException(field + " must encode exactly " + expectedBytes + " bytes");
    }
    if (maximumBytes != null && value.length() > maximumBytes * 2) {
      throw new IllegalArgumentException(field + " exceeds the encoded byte limit");
    }
    return value;
  }

  private static String requireAliasText(final String value, final String field) {
    if (value == null
        || value.isEmpty()
        || !value.equals(value.trim())
        || value.getBytes(StandardCharsets.UTF_8).length > 128) {
      throw new IllegalArgumentException(field + " must be unpadded UTF-8 of at most 128 bytes");
    }
    for (int index = 0; index < value.length(); index++) {
      if (Character.isISOControl(value.charAt(index))) {
        throw new IllegalArgumentException(field + " must not contain control characters");
      }
    }
    return value;
  }

  private static String require(final Map<String, String> arguments, final String key) {
    final String value = arguments.get(key);
    if (value == null || value.isEmpty()) {
      throw new IllegalArgumentException("Instruction argument '" + key + "' is required");
    }
    return value;
  }

  private static long requireLong(final Map<String, String> arguments, final String key) {
    final String value = require(arguments, key);
    try {
      return Long.parseLong(value);
    } catch (final NumberFormatException ex) {
      throw new IllegalArgumentException(
          "Instruction argument '" + key + "' must be a number: " + value, ex);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/RegisterPinManifestInstruction.kt
@file:OptIn(ExperimentalEncodingApi::class)

package org.hyperledger.iroha.sdk.core.model.instructions

import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

private const val SORAFS_REGISTER_PIN_ACTION = "RegisterPinManifest"
private const val MAX_MANIFEST_PAYLOAD_BYTES = 512 * 1024
private const val MAX_ALIAS_PROOF_BYTES = 1024 * 1024
private val canonicalSorafsHex = Regex("^[0-9a-f]+$")

private fun requireCanonicalHex(
    value: String?,
    fieldName: String,
    expectedBytes: Int? = null,
    maximumBytes: Int? = null,
): String {
    require(!value.isNullOrEmpty()) { "$fieldName must not be empty" }
    require(value.length % 2 == 0 && canonicalSorafsHex.matches(value)) {
        "$fieldName must be canonical lowercase even-length hex without a prefix"
    }
    expectedBytes?.let {
        require(value.length == it * 2) { "$fieldName must encode exactly $it bytes" }
    }
    maximumBytes?.let {
        require(value.length <= it * 2) { "$fieldName must encode at most $it bytes" }
    }
    return value
}

private fun requireNonzeroDigest(value: String?, fieldName: String): String {
    val digest = requireCanonicalHex(value, fieldName, expectedBytes = 32)
    require(digest.any { it != '0' }) { "$fieldName must not be the all-zero digest" }
    return digest
}

private fun requireCanonicalManifestPayload(value: String?): String {
    require(!value.isNullOrEmpty()) { "manifestPayloadBase64 must not be empty" }
    val maximumEncodedLength = ((MAX_MANIFEST_PAYLOAD_BYTES + 2) / 3) * 4
    require(value.length <= maximumEncodedLength) {
        "manifestPayloadBase64 exceeds the $MAX_MANIFEST_PAYLOAD_BYTES-byte limit"
    }
    val decoded = try {
        Base64.decode(value)
    } catch (ex: IllegalArgumentException) {
        throw IllegalArgumentException("manifestPayloadBase64 must be valid base64", ex)
    }
    require(decoded.isNotEmpty() && decoded.size <= MAX_MANIFEST_PAYLOAD_BYTES) {
        "manifestPayloadBase64 must encode 1..$MAX_MANIFEST_PAYLOAD_BYTES bytes"
    }
    require(Base64.encode(decoded) == value) {
        "manifestPayloadBase64 must use canonical padded base64"
    }
    return value
}

private fun requireAliasText(value: String?, fieldName: String): String {
    require(!value.isNullOrEmpty()) { "$fieldName must not be empty" }
    require(value == value.trim() && value.toByteArray(Charsets.UTF_8).size <= 128) {
        "$fieldName must be unpadded UTF-8 of at most 128 bytes"
    }
    require(value.none(Char::isISOControl)) { "$fieldName must not contain control characters" }
    return value
}

/** Typed first-release builder for the consensus `RegisterPinManifest` instruction. */
class RegisterPinManifestInstruction private constructor(
    @JvmField val manifestPayloadBase64: String,
    @JvmField val chunkDigestSha3Hex: String,
    @JvmField val submittedEpoch: Long,
    @JvmField val successorOfHex: String?,
    @JvmField val aliasBinding: AliasBinding?,
    override val arguments: Map<String, String>,
) : InstructionTemplate {

    override val kind: InstructionKind get() = InstructionKind.REGISTER

    /** Return a new array containing the canonical Norito manifest payload. */
    fun manifestPayloadBytes(): ByteArray = Base64.decode(manifestPayloadBase64)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RegisterPinManifestInstruction) return false
        return manifestPayloadBase64 == other.manifestPayloadBase64 &&
            chunkDigestSha3Hex == other.chunkDigestSha3Hex &&
            submittedEpoch == other.submittedEpoch &&
            successorOfHex == other.successorOfHex &&
            aliasBinding == other.aliasBinding
    }

    override fun hashCode(): Int {
        var result = manifestPayloadBase64.hashCode()
        result = 31 * result + chunkDigestSha3Hex.hashCode()
        result = 31 * result + submittedEpoch.hashCode()
        result = 31 * result + (successorOfHex?.hashCode() ?: 0)
        result = 31 * result + (aliasBinding?.hashCode() ?: 0)
        return result
    }

    /** Builder for a canonical register-pin instruction. */
    class Builder internal constructor() {
        private var manifestPayloadBase64: String? = null
        private var chunkDigestSha3Hex: String? = null
        private var submittedEpoch: Long? = null
        private var successorOfHex: String? = null
        private var aliasBinding: AliasBinding? = null

        fun setManifestPayloadBase64(manifestPayloadBase64: String) = apply {
            this.manifestPayloadBase64 = requireCanonicalManifestPayload(manifestPayloadBase64)
        }

        fun setManifestPayload(manifestPayload: ByteArray) = apply {
            require(manifestPayload.isNotEmpty() && manifestPayload.size <= MAX_MANIFEST_PAYLOAD_BYTES) {
                "manifestPayload must contain 1..$MAX_MANIFEST_PAYLOAD_BYTES bytes"
            }
            this.manifestPayloadBase64 = Base64.encode(manifestPayload.copyOf())
        }

        fun setChunkDigestSha3Hex(chunkDigestSha3Hex: String) = apply {
            this.chunkDigestSha3Hex = requireNonzeroDigest(chunkDigestSha3Hex, "chunkDigestSha3Hex")
        }

        fun setSubmittedEpoch(submittedEpoch: Long) = apply {
            require(submittedEpoch >= 0) { "submittedEpoch must be non-negative" }
            this.submittedEpoch = submittedEpoch
        }

        fun setSuccessorOfHex(successorOfHex: String?) = apply {
            this.successorOfHex = successorOfHex?.let {
                requireNonzeroDigest(it, "successorOfHex")
            }
        }

        fun setAliasBinding(aliasBinding: AliasBinding?) = apply {
            this.aliasBinding = aliasBinding
        }

        fun build(): RegisterPinManifestInstruction {
            val payload = checkNotNull(manifestPayloadBase64) { "manifestPayload must be set" }
            val chunkDigest = checkNotNull(chunkDigestSha3Hex) { "chunkDigestSha3Hex must be set" }
            val epoch = checkNotNull(submittedEpoch) { "submittedEpoch must be set" }
            val args = canonicalArguments(
                payload,
                chunkDigest,
                epoch,
                successorOfHex,
                aliasBinding,
            )
            return RegisterPinManifestInstruction(
                payload,
                chunkDigest,
                epoch,
                successorOfHex,
                aliasBinding,
                args,
            )
        }
    }

    /** Optional manifest alias binding. */
    class AliasBinding private constructor(
        @JvmField val name: String,
        @JvmField val namespace: String,
        @JvmField val proofHex: String,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is AliasBinding) return false
            return name == other.name && namespace == other.namespace && proofHex == other.proofHex
        }

        override fun hashCode(): Int {
            var result = name.hashCode()
            result = 31 * result + namespace.hashCode()
            result = 31 * result + proofHex.hashCode()
            return result
        }

        internal fun appendArguments(target: MutableMap<String, String>) {
            target["alias.name"] = name
            target["alias.namespace"] = namespace
            target["alias.proof_hex"] = proofHex
        }

        class Builder internal constructor() {
            private var name: String? = null
            private var namespace: String? = null
            private var proofHex: String? = null

            fun setName(name: String) = apply {
                this.name = requireAliasText(name, "alias.name")
            }

            fun setNamespace(namespace: String) = apply {
                this.namespace = requireAliasText(namespace, "alias.namespace")
            }

            fun setProofHex(proofHex: String) = apply {
                this.proofHex = requireCanonicalHex(
                    proofHex,
                    "alias.proofHex",
                    maximumBytes = MAX_ALIAS_PROOF_BYTES,
                )
            }

            fun build(): AliasBinding = AliasBinding(
                checkNotNull(name) { "alias.name must be set" },
                checkNotNull(namespace) { "alias.namespace must be set" },
                checkNotNull(proofHex) { "alias.proofHex must be set" },
            )
        }

        companion object {
            @JvmStatic
            fun builder(): Builder = Builder()

            internal fun fromArguments(arguments: Map<String, String>): AliasBinding? {
                val keys = setOf("alias.name", "alias.namespace", "alias.proof_hex")
                val present = keys.count(arguments::containsKey)
                require(present == 0 || present == keys.size) {
                    "alias binding requires alias.name, alias.namespace, and alias.proof_hex together"
                }
                if (present == 0) return null
                return builder()
                    .setName(requireArgument(arguments, "alias.name"))
                    .setNamespace(requireArgument(arguments, "alias.namespace"))
                    .setProofHex(requireArgument(arguments, "alias.proof_hex"))
                    .build()
            }
        }
    }

    companion object {
        const val ACTION: String = SORAFS_REGISTER_PIN_ACTION

        private val mandatoryArgumentKeys = setOf(
            "action",
            "manifest_payload_base64",
            "chunk_digest_sha3_256_hex",
            "submitted_epoch",
        )
        private val optionalArgumentKeys = setOf(
            "successor_of_hex",
            "alias.name",
            "alias.namespace",
            "alias.proof_hex",
        )

        @JvmStatic
        fun builder(): Builder = Builder()

        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): RegisterPinManifestInstruction {
            require(arguments["action"] == ACTION) { "Instruction argument 'action' must be $ACTION" }
            require(arguments.keys.all { it in mandatoryArgumentKeys || it in optionalArgumentKeys }) {
                "RegisterPinManifest arguments contain unsupported fields"
            }
            require(mandatoryArgumentKeys.all(arguments::containsKey)) {
                "RegisterPinManifest arguments are missing required fields"
            }
            return builder()
                .setManifestPayloadBase64(requireArgument(arguments, "manifest_payload_base64"))
                .setChunkDigestSha3Hex(requireArgument(arguments, "chunk_digest_sha3_256_hex"))
                .setSubmittedEpoch(requireLong(arguments, "submitted_epoch"))
                .setSuccessorOfHex(arguments["successor_of_hex"])
                .setAliasBinding(AliasBinding.fromArguments(arguments))
                .build()
        }

        private fun canonicalArguments(
            manifestPayloadBase64: String,
            chunkDigestSha3Hex: String,
            submittedEpoch: Long,
            successorOfHex: String?,
            aliasBinding: AliasBinding?,
        ): Map<String, String> = buildMap {
            put("action", ACTION)
            put("manifest_payload_base64", manifestPayloadBase64)
            put("chunk_digest_sha3_256_hex", chunkDigestSha3Hex)
            put("submitted_epoch", submittedEpoch.toString())
            successorOfHex?.let { put("successor_of_hex", it) }
            aliasBinding?.appendArguments(this)
        }

        private fun requireArgument(arguments: Map<String, String>, key: String): String {
            val value = arguments[key]
            require(!value.isNullOrEmpty()) { "Instruction argument '$key' is required" }
            return value
        }

        private fun requireLong(arguments: Map<String, String>, key: String): Long {
            val value = requireArgument(arguments, key)
            return try {
                value.toLong()
            } catch (ex: NumberFormatException) {
                throw IllegalArgumentException("Instruction argument '$key' must be a number: $value", ex)
            }
        }
    }
}
